### PR TITLE
TLS 1.3 Introduce and use key exchange mode config options

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
The purpose of this PR is to move away from using in TLS 1.3 the TLS 1.2 specific key exchange configuration options to enable/disable the support of PSK, certificates and ephemeral keys during a TLS 1.3 handshake. TLS 1.2 specific key exchange configuration options (MBEDTLS_KEY_EXCHANGE_*_ENABLED macros) are associated to sets of TLS 1.2 ciphersuites. TLS 1.2 ciphersuites do not fit well with TLS 1.3 where the key exchange mechanism during an handshake is not defined by the ciphersuite in use.

This PR introduces and uses key exchange mode config options that map the three TLS 1.3 key exchange mode: psk, ephemeral and psk_ephemeral. This makes the TLS 1.3 implementation easier to maintain and evolve.

Before to introduce the new TLS 1.3 key exchange mode config options, this PR fixes issues that where found when testing builds without all TLS 1.3 key exchange modes enabled.

Preparatory work for #6379.

## Status
IN REVIEW

## Requires Backporting
NO, no support for TLS 1.3 in 2.28.

## Migrations
NO

### CODE REVIEW

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Except document typo issues, other part looks good to me. 
I post some minor improvements.  Please check if it should be change in this PR.

File: docs/architecture/tls13-support.md [hunk 1]
```diff
@@ -141,10 +141,33 @@ Support description
...
-      of the MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_NONE/PSK/PSK_EPHEMERAL/EPHEMERAL
-      runtime configuration macros.
+  (2) See the TLS 1.3 specific build options section below.
+
+- TLS 1.3 specific build options:
+
+  - MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE enables the support for middlebox
+    compatibility mode as defined in section D.4 of RFC 8446.
+
+  - MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED enables the
```
Comment (xkqian) [inline docs/architecture/tls13-support.md:151]:
```suggestion
  - MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED enables the
```

**Review by xkqian — CHANGES_REQUESTED** [review]

File: library/ssl_tls.c [hunk 2]
```diff
@@ -1593,16 +1593,21 @@ int mbedtls_ssl_set_hs_ecjpake_password( mbedtls_ssl_context *ssl,
 }
 #endif /* MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED */
 
-#if defined(MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
-
-MBEDTLS_CHECK_RETURN_CRITICAL
-static int ssl_conf_psk_is_configured( mbedtls_ssl_config const *conf )
+#if defined(MBEDTLS_SSL_HANDSHAKE_WITH_PSK_ENABLED)
+int mbedtls_ssl_conf_has_static_psk( mbedtls_ssl_config const *conf )
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply library/ssl_tls.c:1597]: I prefer it not being inline. Too big for that to me.

File: tests/ssl-opt.sh [hunk 3]
```diff
@@ -251,6 +251,31 @@ requires_any_configs_disabled() {
...
+                                MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED"
+
+requires_key_exchange_with_cert_in_tls12_enabled() {
+    if ! $P_QUERY -any $TLS1_2_KEY_EXCHANGES_WITH_CERT
+    then
+        SKIP_NEXT="YES"
+    fi
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:266]: I've removed `requires_key_exchange_with_cert_in_tls12_enabled` and replaced it by this everywhere.

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -251,6 +251,31 @@ requires_any_configs_disabled() {
...
+}
+
+requires_key_exchange_with_cert_in_tls12_or_tls13_enabled() {
+    if $P_QUERY -all MBEDTLS_SSL_PROTO_TLS1_2
+    then
+        requires_key_exchange_with_cert_in_tls12_enabled
+    elif ! $P_QUERY -all MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
+    then
+        SKIP_NEXT="YES"
+    fi
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:276]: That's not equivalent. If TLS 1.2 is enabled in the build we do not want to take into account MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED.

File: library/ssl_tls13_client.c [hunk 5]
```diff
@@ -1141,15 +1136,17 @@ int mbedtls_ssl_tls13_write_client_hello_exts( mbedtls_ssl_context *ssl,
...
 
+#if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_SOME_EPHEMERAL_ENABLED)
     if( mbedtls_ssl_conf_tls13_some_ephemeral_enabled( ssl ) )
     {
         ret = ssl_tls13_write_key_share_ext( ssl, p, end, &ext_len );
         if( ret != 0 )
             return( ret );
         p += ext_len;
     }
+#endif
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply library/ssl_tls13_client.c:1147]: There is no strict rule. Here I find it is readable enough without.

**Review by yuhaoth — APPROVED** [review]
LGTM

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -251,6 +251,24 @@ requires_any_configs_disabled() {
     fi
 }
 
+TLS1_2_KEY_EXCHANGES_WITH_CERT="MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED \
```
Comment (xkqian) [inline tests/ssl-opt.sh:255]:
```suggestion
```

**Review by xkqian — COMMENTED** [review]
This Looks quite good to me, only some minor improvement needed.

File: docs/architecture/tls13-support.md [hunk 7]
```diff
@@ -141,10 +141,33 @@ Support description
...
+    support of the ephemeral key exchange mode. If it is the only key exchange
+    mode enabled, the TLS 1.3 implementation does not contain any code related
+    to PSK based key exchange. The ephemeral key exchange mode requires at least
+    one of the key exchange protocol allowed by the TLS 1.3 specification, the
+    parsing and validation of x509 certificates and at least one signature
+    algorithm allowed by the TLS 1.3 specification for signature computing and
+    verification.
+
+  - MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED enables the
+    support for the PSK ephemeral key exchange mode. If it is the only key
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply docs/architecture/tls13-support.md:166]: Looking at Oxford Learner's Dictionaries it seems that "for" is the correct one. Thus I've changed the "of" line 157.

File: tests/ssl-opt.sh (hunk 6 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:255]: In fact one of the two should have been `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED`.

**Review by xkqian — APPROVED** [review]
LGTM

**Review by yuhaoth — APPROVED** [review]
LGTM

**Comment by xkqian** [comment]
There are conflict files. Seems need rebase

**Comment by ronald-cron-arm** [comment]
I've rebased. The conflicts were mostly (if not all) due to the merge of #6385. The main impact was on commit "Introduce and use MBEDTLS_SSL_HANDSHAKE_WITH_CERT_ENABLED" due to the addition of many "#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)" guards in ssl_client2.c, ssl_server2.c and test_suite_ssl.function.

**Review by xkqian — APPROVED** [review]
LGTM

File: tests/ssl-opt.sh [hunk 8]
```diff
@@ -251,6 +251,24 @@ requires_any_configs_disabled() {
...
 
+TLS1_2_KEY_EXCHANGES_WITH_CERT="MBEDTLS_KEY_EXCHANGE_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED \
+                                MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED"
+
+requires_key_exchange_with_cert_in_tls12_or_tls13_enabled() {
```
Comment (gilles-peskine-arm) [inline tests/ssl-opt.sh:262]: Is this supposed to be a corrected implementation of `requires_certificate_authentication`, or does it have different intended semantics?

I'm surprised that this pull requests adds a lot of explicit calls to `requires_key_exchange_with_cert_in_tls12_or_tls13_enabled`. We already have quite a bit of automatic detection of certificate requirements. Do you think these explicit calls would be needed if `requires_certificate_authentication` was fixed to handle TLS 1.3 and builds with PSK-ephemeral but no cert?

**Pull request merged**